feat(adr): add Service Registry ADR

tonyespy opened the pull request:
Signed-off-by: Tony Espy <espy@canonical.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Changes have been rendered and validated locally using mkdocs-material (see edgex-docs README)
- [x] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-docs/blob/master/.github/Contributing.md.

## What is the current behavior?
This the initial draft PR for an ADR which discusses the existing behavior of EdgeX microservices with respect to the registry and associated client configuration read from configuration.toml files.

## Issue Number:
NA

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,106 @@
+./internal/system/agent/direct/metrics.go:		e, err := m.registryClient.GetServiceEndpoint(serviceName)
+```
+
+**Note** - I haven't yet bisected the code to determine when (or if) this behavior was changed. In the first (and only)
````
Correct, this was the only location that was calling `GetServiceEndpoint`

tonyespy replied:
I've updated the ADR, and further described the history, so marking this resolved.

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,106 @@
+        GetServiceEndpoint(serviceId string) (types.ServiceEndpoint, error)
+
+        // Checks with the Registry if the target service is available, i.e. registered and healthy
+        IsServiceAvailable(serviceId string) (bool, error)
```
Device services use IsServiceAvailable() on start-up to check if Core Data and Core Metadata are available, but don't request the endpoint.

lenny-intel replied:
Also note that in Geneva and Hanoi the Service health checks registered by the services are used by the Consul scripts to stage gate security bootstrapping.

tonyespy replied:
I've updated the ADR to mention this, so marking this convesation resolved.

jpwhitemn commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,113 @@
+
+This ADR is meant to address the current usage of the registry by EdgeX services, and in particular whether the EdgeX services
+are using the registry (when enabled) to determine the location of peer services vs. using static per-service configuration.
+The reason this is being investigated is that there's been a proposal that EdgeX do away the registry functionality, as the
```
do away -> do away with

jpwhitemn commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,113 @@
+According to the original Service Name Design document (v6)[https://wiki.edgexfoundry.org/display/FA/Architecture+Issues+and+Decisions?preview=%2F7602423%2F12124493%2FServiceNameDesign-v6.pdf], all EdgeX Foundry microservices should be able to accomplish the following tasks:
+
+ * Register with the configuration/registration (referred to simply as “the registry” for the rest of this document) provider (today Consul)
+ * Respond to shutdown requestsoTaking care to clean up of any resources in an orderly fashionoUnregister itself from the registry
```
requestsoTaking -> requests. Taking

jpwhitemn commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,113 @@
+According to the original Service Name Design document (v6)[https://wiki.edgexfoundry.org/display/FA/Architecture+Issues+and+Decisions?preview=%2F7602423%2F12124493%2FServiceNameDesign-v6.pdf], all EdgeX Foundry microservices should be able to accomplish the following tasks:
+
+ * Register with the configuration/registration (referred to simply as “the registry” for the rest of this document) provider (today Consul)
+ * Respond to shutdown requestsoTaking care to clean up of any resources in an orderly fashionoUnregister itself from the registry
```
fashionUnregister -> fashion. Unregister

tonyespy replied:
ADR updated

jpwhitemn commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,113 @@
+
+Next I followed the same steps, but instead of completely removing the client, I instead set the configuration items in this subsection
+to empty strings or 0 in the case of Port. The same error as above occured, and the service exited.
+
```
Is this functionality baked into the SDK or just the virtual service?  Is the current state of the C device services and SDK the same?

tonyespy replied:
This functionality is implemented in device-sdk-go (see [internal/clients/init.go](https://github.com/edgexfoundry/device-sdk-go/blob/93569aefbcf2dcd4d18a31fcf2c190494ae30d88/internal/clients/init.go#L54)).

As for device-sdk-c, I took a cursory look, and it appears it actually does query the registry for the host and port of core data and metadata (see [src/c/service.c](https://github.com/edgexfoundry/device-sdk-c/blob/master/src/c/service.c#L730)) if the registry flag has been set, otherwise it uses the clients from `configuration.toml`. @iain-anderson can you please confirm?

bnevis-i commented:
Looks like a good start on this PR.  I would like the PR to include some mention of extra metadata required for services and how that is handled by the registry.  Just knowing host and port doesn't tell the caller whether or not that port is speaking UDP, TCP, AMQP, SMTP, HTTP, HTTP over TLS, or the database name for a database.

bnevis-i commented:
@tonyespy Awaiting updates to the ADR.

bnevis-i commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,183 @@
+## Decision
+Update the core, support, and security-proxy-setup to use go-mod-registry's ```Client.GetServiceEndpoint``` method on startup to determine (a) if
+a service dependency is available and (b) use the returned address information to initialize client endpoints (or setup the correct route in the
+case of proxy-setup).
````
What about additional metadata such as:
- The URL path described earlier
- Which protocol to use to connect (tcp, http, https, etc...)

Where did this metadata go.  It used to be part of the original spec:
```
MetaDeviceURL = "http://edgex-core-metadata:48081/api/v1/device"
```

tonyespy replied:
The URL is constructed using a set of constants which define the available endpoints. In the original Java code (and the first version of the re-written Go services) the actual endpoint paths were defined in the configuration files which made no sense at all.  Just because an endpoint path was added to a configuration file, didn't mean the service being called actually implements the given endpoint.

Also with regards to protocol, all of our endpoints only support HTTP. For instances where additional protocols are supported, additional configuration is required.

tonyespy replied:
Marking this resolved. If there's still an issue, feel free to re-open with relevant details.

bnevis-i commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,183 @@
+```
+
+### Security Proxy Setup
+The security-proxy-setup service also relies on static service address configuration to configure the server routes for each
+of the services accessible through the API Gateway (aka Kong). Although it uses the same TOML-based client config keys as the
+other services, these configuration values are only ever read from the security-proxy-setup's local configuration.toml file,
+as the security services have never supported using our configuration provider (aka Consul).
````
What is the proposal on what security proxy setup should do?

If security proxy setup uses the registry, and services haven't registered by the time that proxy setup runs, what then?

tonyespy replied:
Well... I imagine the only way this could be done cleanly would be to support lazy initialization. Let's discuss at tomorrow's WG meeting.

bnevis-i reviewed (commented):
Dismissing disapproving review.

tonyespy commented:
As there's been at least one approval review submitted, I'm taking this PR out of draft status. Also I pushed one additional fixup commit due to @bnevis-i's comment about security-proxy-setup.

tonyespy commented:
I also rebased to sync with master and re-pushed.

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,192 @@
+level=ERROR app=device-virtual source=init.go:45 msg="Couldn't register to metadata service: Get \"http://localhost:3112/api/v1/deviceservice/name/device-virtual\": dial tcp 127.0.0.1:3112: connect: connection refused\n"
+```
+
+**Note** - in order to run this second test, the easiest way to do so is to remove remove and reinstall the snap vs. manually wiping
````
Change `remove remove and reinstall` to `remove and reinstall`

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,192 @@
+filed (https://github.com/edgexfoundry/edgex-go/issues/2594), and the client monitoring was removed as part of the Geneva 1.2.1 release.
+
+## Problem Statement
+The fundamental problem with the existing implementations (as decribed above), is that there is too much duplication of configuration across
+services. For instance, Core Data's service port can easily be changed by passing the environment variable SERVICE_PORT to the service on
+startup. This overrides the configuration read from the configuration provider, and will cause Core Data to listen on the new port,
+however it has no impact on any services which use Core Data, as the client config for each is read from the configuration provider (excluding
+security-proxy-setup).
+
+This means in order to change a service port, environment variable overrides need to set for every client service as well as security-proxy-setup
+(if required).
```
For this to work the `CLIENTS_COREDATA_PORT` environment variable would also need to be changed for each service. In docker we add it to the common.env file and re-generate the the compose files, which sets the override on each service. For snap is there a single environment for all the service or does each have it own like in docker?

tonyespy replied:
Mentioned `CLIENTS_COREDATA_PORT` in the text.

As for whether snaps have the ability to define a single environment variable for all services...  Yes, environment variables can be defined at buildtime (i.e. in snapcraft.yaml) which apply to all services. At runtime, this is up to the snap developer. In fact this is actually implemented in the edgexfoundry snap for `EDGEX_SECURITY_SECRET_STORE` variable. So in theory, it *could* also be implemented for client addressing if/when the registry was not in use.

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,192 @@
+(if required).
+
+## Decision
+Update the core, support, and security-proxy-setup to use go-mod-registry's ```Client.GetServiceEndpoint``` method on startup to determine (a) if
````
The App and Device SDKs need to also be included.

tonyespy replied:
addressed in latest update

bnevis-i reviewed (changes requested):
The proposal calls for everyone to now call `GetServiceEndpoint()` to get host and port information for remote services?  How does `GetServiceEndpoint()` behave when the `-registry` flag hasn't been passed?

tonyespy commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,183 @@
+While the SDK does use the ```IsServiceAvailable``` method on startup to confirm that its dependencies are available, it uses client config
+keys to initialize its various REST API endpoints used to interact with Core Data and Metadata.
+
+**NOTE** - the behavior of device-sdk-c has not yet been verified.
````
Summary added which covers all services in edgex-go, as well as App Functions SDK, and both Device SDKs. As such, marking this conversation resolved.

tonyespy commented:
> The proposal calls for everyone to now call `GetServiceEndpoint()` to get host and port information for remote services? How does `GetServiceEndpoint()` behave when the `-registry` flag hasn't been passed?

The static configuration is used. See previous description of the C Device SDK, as well as the original design document.

bnevis-i commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,200 @@
+a service dependency is available and (b) use the returned address information to initialize client endpoints (or setup the correct route in the
+case of proxy-setup). The same changes also need to be applied to the App Functions SDK and Go Device SDK, with only minor changes required in the C Device SDK (see previous commments re: the current implementation).
+
+*Note* - One impact of this decision is that as the security-proxy-setup service currently runs _before_ any of the core and support services are
+started, it would not be possible to implement this proposal without also modifying the service to use a lazy initialization of the API Gateway's
+routes. We should also explore whether it's possible to extend the registry abstraction to allow notification of new service registrations vs. forcing
+proxy-setup to poll for all known services.
```
IMO this caveat about security-proxy-setup is fatal to applying this ADR to this service.  We need to define a mechanism defined that will keep security-proxy-setup from doing its thing until the required services are running, or figure out how to make proxy-setup dynamic in some way.  If we go ahead with this ADR as written proxy-setup will act non-deterministically.

beaufrusetta replied:
Define "lazy initialization". 

[**n00b dev**]  _**I may be totally wrong here**_  [/**n00b dev**], but it is my understanding that proxy setup doesn't wait for services to be up in order to create Kong services/routes/plugins - it just configures them through the Kong Admin API and exits. From my limited knowledge, the bootstrapping process proxy setup just waits on the Kong service and Kong DB to be up and running, and then the subsequent Go code starts executing. Yeah/no?

bnevis-i replied:
> [**n00b dev**] _**I may be totally wrong here**_ [/**n00b dev**], but it is my understanding that proxy setup doesn't wait for services to be up in order to create Kong services/routes/plugins - it just configures them through the Kong Admin API and exits.

There is a fine line between a service REGISTERING ITSELF and the service BEING UP.  You're right.  Proxy setup doesn't require the service to be up.  But Tony's ADR requires the service to be REGISTERED in order to obtain the address and port for configuration.  The line is fine because the they between a service registering itself and being able to serve requests is only milliseconds.

So if we want to rely on the registry, we have a few options.

## Option 1

In Consul mode, proxy setup would use not its full compiled-in service registry, but simply a LIST OF REQUIRED SERVICES.  Proxy setup would either have to wait for all those services to have registered before configuring the proxy, or will need to be a long-running services that watches for new service registrations and adds them to them to the proxy if they appear in the list.

I am not aware of any use cases where Kong is used without Consul.


## Option 2

Leverage the ability to [attach metadata to service registrations](https://www.consul.io/api-docs/catalog#register-entity) and have each service add metadata that indicates that it wants to be exposed via the Kong proxy.  Proxy-setup could be a daemon process that, as above, watched for new service registrations with the flag set.  We could also encode metadata such as what ACL group(s?) the service belongs to, so that one could have finer-grained controls as to which services are available with a given Kong token.


The point I was trying to make above (the "fatal" bit), is that we can't hand-wave it.  It needs to be defined in the ADR.

lenny-intel replied:
I think **Option 2** is doable. We can add a new property something  like `ExposeViaProxy` to the `RegistryInfo` configuration and update go-mod-register to add this to the metadata for service registry. The Registry config section would look like:
```
[Registry]
Host = 'localhost'
Port = 8500
Type = 'consul'
ExposeViaProxy = true
```
We would also expand the the Registry's interface to add an API to return all registered services and their metadata with optional metadata filter. 

I looked at the Consul API and verified that we can do the above.

proxy-setup would then be long lived adding proxy for any new services that are registered. Downside is it will have to poll the list of registered services periodically. 

 I looks like the same QueryOptions we use to wait for config changes exist on the API to get a Service's info (i.e. endpoint), but not on the API that returns all the registered services. This means that the Clients could use this to refresh the endpoint if it changes.

jpwhitemn commented:
Per Security WG meeting 2/17/21 - discussion between Bryon and Tony (regarding proxy setup and potential artificial dependency that doesn't need to be there), re-review by Lenny remain on this.

Does Kong have a feature to pull this out of Consul directly?  I.e. - Proxy everything you have in Consul automatically - changing this slightly with regard to proxy setup (maybe allowing proxy setup to go away).  Get with Colin to clarify.

Target to get this approved before Ireland.  Implementation TBD - but because it is backward compatible, could be done in Jakarta.

bnevis-i commented:
This ADR number overlaps the ADR number used in #268

tonyespy commented:
I just updated the consequences section of the ADR based upon the discussion in the security working group meeting on Feb 24 '21. This should be in good shape for final approval...

bnevis-i commented on docs_src/design/adr/0006-Service-Registry.md:
```diff
@@ -0,0 +1,200 @@
+however it has no impact on any services which use Core Data, as the client config for each is read from the configuration provider (excluding
+security-proxy-setup).
+
+This means in order to change a service port, environment variable overrides (e.g. CLIENTS_COREDARA_PORT) need to set for every client service as well as security-proxy-setup (if required).
```
Lenny's comment above (typo fixed) still not addressed.

tonyespy commented:
I just updated the consequences and decisions sections of the ADR based upon the discussion in the security working group meeting on Mar 10 '21. This should be in good shape for final approval...

Note, I force pushed as I rebased against master.

lenny-intel commented on docs_src/design/adr/0006-Service-Registry.md:
````diff
@@ -0,0 +1,200 @@
+
+## Decision
+Update the core, support, and security-proxy-setup services to use go-mod-registry's ```Client.GetServiceEndpoint``` method on startup to determine (a) if
+a service dependency is available and (b) use the returned address information to initialize client endpoints (or setup the correct route in the
````
@tonyespy , add this note here or consequence's. It is important this is understood that we have to make sure these occur in proper order.

tonyespy replied:
Just pushed one more commit which addresses these comments...

tonyespy commented:
> This ADR number overlaps the ADR number used in #268

Yes it does. I seem to recall in an earlier ADR review I asked the question of how we allocate ADR numbers, and I don't ever think I saw a reply.

@jpwhitemn @lenny-intel is there a process for this documented anywhere that I somehow missed?

lenny-intel reviewed (approved):
LGTM

bnevis-i reviewed (approved):
LGTM, repeating comment that PR 268 (filed before this one) is already trying to grab ADR ID 0006.

jpwhitemn commented:
Approved by TSC 3/25/21
